### PR TITLE
Migration

### PR DESCRIPTION
**Pull request by cl-shaharshavit-rnd**

### CODE REVIEW

File: src/multer/storage/fileStstemStorageBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  public createStorage(): multer.StorageEngine {
+    return multer.diskStorage({
+      destination: this.createUploadFolder.bind(this),
+      filename: (req: Request, file: Express.Multer.File, cb: Callback) => {
+        cb(null, file.originalname);
+      },
+    });
+  }
+  private createUploadFolder(
```
Comment (asafMasa) [inline src/multer/storage/fileStstemStorageBuilder.ts:23]: add a row space

**Review by asafMasa — CHANGES_REQUESTED** [review]
great job, and you are missing a lot of logs

File: confd/production.tmpl [hunk 2]
```diff
@@ -5,5 +5,25 @@
...
+      "signatureVersion": "{{ getv "/configuration/storage/s3/signatureVersion" "v4" }}",
+      "region": "{{ getv "/configuration/storage/s3/region" "" }}",
+      "apiVersion": "{{ getv "/configuration/storage/s3/apiVersion" "2006-03-01" }}"
+    },
+    "FS": {
+      "uploadRoot": "{{ getv "/configuration/storage/fs/uploadRoot" "D:/tmp" }}"
+    }
+  },
+  "dependentServices": {
+    "imageIndexerBaseUrl": "{{ getv "/configuration/dependentServices/imageIndexerBaseUrl" "http://localhost:3001/images" }}",
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply confd/production.tmpl:26]: we use to to check if the file exists to block create request that will override data and update requests with missing data

File: src/containerConfig.ts [hunk 3]
```diff
@@ -16,6 +19,25 @@ function registerExternalValues(): void {
...
+
+  registerStorageEngine(logger);
+}
+
+function registerStorageEngine(logger: MCLogger): void{
+  const storageEngine = get<string>('storage.engine');
+  let storage: StorageEngine;
+  switch (storageEngine.toUpperCase()) {
+    case 'FS':
+      storage = new FileSystemStorageBuilder(logger).createStorage();
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/containerConfig.ts:31]: linter wont let me create variables in switch case

File: src/containerConfig.ts [hunk 4]
```diff
@@ -16,6 +19,25 @@ function registerExternalValues(): void {
...
+      break;
+    case 'S3':
+      storage = new S3StorageBuilder(logger).createStorage();
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/containerConfig.ts:34]: linter wont let me create variables in switch case

File: src/models/createUploadRequest.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { ImageMetadata } from '@map-colonies/mc-model-types';
+
+export interface CreateUploadRequest {
+  additionalData: ImageMetadata;
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/models/createUploadRequest.ts:4]: that was the definitions when we made the service.
it should be changed to match current epic but it shouldn't be part of this PR

File: src/services/catalogDbService.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    return data.exists;
+  }
+
+  public async getById(id: string): Promise<ImageMetadata> {
+    const url = joinUrl(this.baseUrl, id);
+    this.logger.debug('sending http get request to: %s', url);
+    const res = await axios.get<ApiHttpResponse>(url).catch((err)=>{
+      const error = err as Error;
+      this.logger.error(`getById request to ${url} failed: ${error.message}`);
+      this.logger.debug(`trace: ${error.stack as string}`);
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/services/catalogDbService.ts:35]: look at the log levels

File: src/services/catalogDbService.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  public constructor(@inject(delay(() => MCLogger)) private readonly logger: MCLogger) {
+    this.baseUrl = config.get('dependentServices.imageIndexerBaseUrl');
+  }
+
+  public async exists(id: string): Promise<boolean> {
+    const url = joinUrl(this.baseUrl, 'exists', id);
+    this.logger.debug('sending http get request to: %s', url);
+    const res = await axios.get<ApiHttpResponse>(url).catch((err)=>{
+      const error = err as Error;
+      this.logger.error(`exsists request to ${url} failed: ${error.message}`);
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/services/catalogDbService.ts:21]: look at the log levels

File: src/multer/storage/fileStstemStorageBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import { mkdirs } from 'fs-extra';
+import { MCLogger } from '@map-colonies/mc-logger';
+import { Request } from 'express';
+import config from 'config';
+import { CreateUploadRequest } from '../../models/createUploadRequest';
+import { UpdateUploadRequest } from '../../models/updateUploadRequest';
+
+declare type Callback = (error: Error | null, destination: string) => void;
+
+export class FileSystemStorageBuilder {
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/multer/storage/fileStstemStorageBuilder.ts:12]: rename to what?

File: src/multer/storage/fileStstemStorageBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import { normalize, join } from 'upath';
+import * as multer from 'multer';
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/multer/storage/fileStstemStorageBuilder.ts:2]: didn't try this time but when we created the original version this it didn't work with any other way  (something with the typescript compilation)

File: src/multer/storage/fileStstemStorageBuilder.ts (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply src/multer/storage/fileStstemStorageBuilder.ts:23]: fixed

File: src/containerConfig.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/containerConfig.ts:31]: as discussed

File: src/containerConfig.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/containerConfig.ts:34]: as discussed

File: src/models/createUploadRequest.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/models/createUploadRequest.ts:4]: ok, open an issue on this and then we will resolve this comment

File: src/multer/storage/fileStstemStorageBuilder.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/multer/storage/fileStstemStorageBuilder.ts:12]: as discussed

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/models/createUploadRequest.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply src/models/createUploadRequest.ts:4]: https://github.com/MapColonies/catalog-upload-service/issues/5

**Review by asafMasa — APPROVED** [review]

**Pull request merged**